### ISSUE TITLE
RedissonLock - infinity loop of acquiring/releasing, the number of available permits is greater than 1.

### ISSUE THREAD

**Issue by adr-wojtasik** [redisson/redisson#6025]
### Redis version

7.2.3

### Redisson version

3.32.0

### Redisson configuration

Single server with default settings.

### What is the Expected behavior?

The `RedissonLock` should work properly and the number of available permits in the lock should be always 0 or 1, it never should be greater.

### What is the Actual behavior?

Due to network issues(?), the locking mechanism starts working in the wrong way. The number of permits held by lock increases unexpectedly above 1. In this case, the publisher thread falls into an infinite loop of acquiring/releasing. 

## Issue
Some time ago we decided to switch from the `RedissonBlockingQueue` to the `RedissonPriorityBlockingQueue`. The application has worked stable for 2-3 days and unexpectedly stopped working. After a deeper investigation, we have noticed that the problem is related to the `RedissonLock` and its locking mechanism. The publisher thread falls into an infinite loop of acquiring/releasing. This happens because the number of available permits held increases unexpectedly above 1. The lock is never removed and publish is never executed.

## Errors in the environment
Our application has 1 publisher thread and 5 consumer threads. 
In the logs from the environment we occasionally see the following errors which might be somehow related to the mentioned issue:

```java
org.redisson.client.RedisResponseTimeoutException: Redis server response timeout (3000 ms) occured after 0 retry attempts, is non-idempotent command: true Check connection with Redis node: ... for TCP packet drops or bandwidth limits.  Try to increase nettyThreads and/or timeout settings. Command: (LPOP), params: [INBOUND], channel: [id: 0x259f3fc0, ...]
	at org.redisson.command.RedisExecutor.lambda$scheduleResponseTimeout$11(RedisExecutor.java:447)
	at io.netty.util.HashedWheelTimer$HashedWheelTimeout.expire(HashedWheelTimer.java:672)
	at io.netty.util.HashedWheelTimer$HashedWheelBucket.expireTimeouts(HashedWheelTimer.java:747)
	at io.netty.util.HashedWheelTimer$Worker.run(HashedWheelTimer.java:472)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Unknown Source)
```
```java
org.redisson.client.RedisTimeoutException: Command execution timeout for command: (PING), params: [], Redis client: [addr=...]
	at org.redisson.client.RedisConnection.lambda$async$0(RedisConnection.java:244)
	at io.netty.util.HashedWheelTimer$HashedWheelTimeout.expire(HashedWheelTimer.java:672)
	at io.netty.util.HashedWheelTimer$HashedWheelBucket.expireTimeouts(HashedWheelTimer.java:747)
	at io.netty.util.HashedWheelTimer$Worker.run(HashedWheelTimer.java:472)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Unknown Source)
```

## Stack trace from the publisher thread
```java
jdk.internal.misc.Unsafe.park(Native Method)
java.util.concurrent.locks.LockSuppor.parkNanos(LockSupport.java:234)
java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedNanos(AbstractQueuedSynchronizer.java:1079)
java.util.concurrent.locks.AbstractQueuedSynchronizer.tryAcquireSharedNanos(AbstractQueuedSynchronizer.java:1369)
org.redisson.RedissonLock.lock(RedissonLock.java:134)
org.redisson.RedissonLock.lock(RedissonLock.java:80)
org.redisson.RedissonPriorityQueue.add(RedissonPriorityQueue.java:156)
```

## Logs from the Redis server
```
1720591766.502783 [0 172.17.0.1:61118] "EVAL" "if ((redis.call('exists', KEYS[1]) == 0) or (redis.call('hexists', KEYS[1], ARGV[2]) == 1)) then redis.call('hincrby', KEYS[1], ARGV[2], 1); redis.call('pexpire', KEYS[1], ARGV[1]); return nil; end; return redis.call('pttl', KEYS[1]);" "1" "redisson_sortedset_lock:{INBOUND}" "30000" "acc3f971-3a53-42a2-b47d-37de483f9e51:46"
1720591766.502841 [0 lua] "exists" "redisson_sortedset_lock:{INBOUND}"
1720591766.502857 [0 lua] "hexists" "redisson_sortedset_lock:{INBOUND}" "acc3f971-3a53-42a2-b47d-37de483f9e51:46"
1720591766.502873 [0 lua] "hincrby" "redisson_sortedset_lock:{INBOUND}" "acc3f971-3a53-42a2-b47d-37de483f9e51:46" "1"
1720591766.502886 [0 lua] "pexpire" "redisson_sortedset_lock:{INBOUND}" "30000"
1720591766.506120 [0 172.17.0.1:61116] "LPOP" "INBOUND"
1720591766.511873 [0 172.17.0.1:61068] "EVAL" "local val = redis.call('get', KEYS[3]); if val ~= false then return tonumber(val);end; if (redis.call('hexists', KEYS[1], ARGV[3]) == 0) then return nil;end; local counter = redis.call('hincrby', KEYS[1], ARGV[3], -1); if (counter > 0) then redis.call('pexpire', KEYS[1], ARGV[2]); redis.call('set', KEYS[3], 0, 'px', ARGV[5]); return 0; else redis.call('del', KEYS[1]); redis.call(ARGV[4], KEYS[2], ARGV[1]); redis.call('set', KEYS[3], 1, 'px', ARGV[5]); return 1; end; " "3" "redisson_sortedset_lock:{INBOUND}" "redisson_lock__channel:redisson_sortedset_lock:{INBOUND}" "redisson_unlock_latch:redisson_sortedset_lock:{INBOUND}:7ec61d55439de3b504ed70e1e1c42cb6" "0" "30000" "acc3f971-3a53-42a2-b47d-37de483f9e51:46" "PUBLISH" "13500"
1720591766.511946 [0 lua] "get" "redisson_unlock_latch:redisson_sortedset_lock:{INBOUND}:7ec61d55439de3b504ed70e1e1c42cb6"
1720591766.511964 [0 lua] "hexists" "redisson_sortedset_lock:{INBOUND}" "acc3f971-3a53-42a2-b47d-37de483f9e51:46"
1720591766.511980 [0 lua] "hincrby" "redisson_sortedset_lock:{INBOUND}" "acc3f971-3a53-42a2-b47d-37de483f9e51:46" "-1"
1720591766.511993 [0 lua] "pexpire" "redisson_sortedset_lock:{INBOUND}" "30000"
1720591766.512008 [0 lua] "set" "redisson_unlock_latch:redisson_sortedset_lock:{INBOUND}:7ec61d55439de3b504ed70e1e1c42cb6" "0" "px" "13500"
1720591766.515069 [0 172.17.0.1:61062] "DEL" "redisson_unlock_latch:redisson_sortedset_lock:{INBOUND}:7ec61d55439de3b504ed70e1e1c42cb6"
1720591767.907611 [0 172.17.0.1:61080] "EVAL" "if ((redis.call('exists', KEYS[1]) == 0) or (redis.call('hexists', KEYS[1], ARGV[2]) == 1)) then redis.call('hincrby', KEYS[1], ARGV[2], 1); redis.call('pexpire', KEYS[1], ARGV[1]); return nil; end; return redis.call('pttl', KEYS[1]);" "1" "redisson_sortedset_lock:{INBOUND}" "30000" "acc3f971-3a53-42a2-b47d-37de483f9e51:46"
1720591767.907661 [0 lua] "exists" "redisson_sortedset_lock:{INBOUND}"
1720591767.907675 [0 lua] "hexists" "redisson_sortedset_lock:{INBOUND}" "acc3f971-3a53-42a2-b47d-37de483f9e51:46"
1720591767.907690 [0 lua] "hincrby" "redisson_sortedset_lock:{INBOUND}" "acc3f971-3a53-42a2-b47d-37de483f9e51:46" "1"
1720591767.907702 [0 lua] "pexpire" "redisson_sortedset_lock:{INBOUND}" "30000"
1720591767.913575 [0 172.17.0.1:61122] "LPOP" "INBOUND"
1720591767.927184 [0 172.17.0.1:61096] "EVAL" "local val = redis.call('get', KEYS[3]); if val ~= false then return tonumber(val);end; if (redis.call('hexists', KEYS[1], ARGV[3]) == 0) then return nil;end; local counter = redis.call('hincrby', KEYS[1], ARGV[3], -1); if (counter > 0) then redis.call('pexpire', KEYS[1], ARGV[2]); redis.call('set', KEYS[3], 0, 'px', ARGV[5]); return 0; else redis.call('del', KEYS[1]); redis.call(ARGV[4], KEYS[2], ARGV[1]); redis.call('set', KEYS[3], 1, 'px', ARGV[5]); return 1; end; " "3" "redisson_sortedset_lock:{INBOUND}" "redisson_lock__channel:redisson_sortedset_lock:{INBOUND}" "redisson_unlock_latch:redisson_sortedset_lock:{INBOUND}:e11bd30c6fcad8656111386e3cf5168a" "0" "30000" "acc3f971-3a53-42a2-b47d-37de483f9e51:46" "PUBLISH" "13500"
1720591767.927242 [0 lua] "get" "redisson_unlock_latch:redisson_sortedset_lock:{INBOUND}:e11bd30c6fcad8656111386e3cf5168a"
1720591767.927258 [0 lua] "hexists" "redisson_sortedset_lock:{INBOUND}" "acc3f971-3a53-42a2-b47d-37de483f9e51:46"
1720591767.927272 [0 lua] "hincrby" "redisson_sortedset_lock:{INBOUND}" "acc3f971-3a53-42a2-b47d-37de483f9e51:46" "-1"
1720591767.927283 [0 lua] "pexpire" "redisson_sortedset_lock:{INBOUND}" "30000"
1720591767.927295 [0 lua] "set" "redisson_unlock_latch:redisson_sortedset_lock:{INBOUND}:e11bd30c6fcad8656111386e3cf5168a" "0" "px" "13500"
1720591767.931136 [0 172.17.0.1:61102] "DEL" "redisson_unlock_latch:redisson_sortedset_lock:{INBOUND}:e11bd30c6fcad8656111386e3cf5168a"
1720591768.803359 [0 172.17.0.1:61090] "EVAL" "if ((redis.call('exists', KEYS[1]) == 0) or (redis.call('hexists', KEYS[1], ARGV[2]) == 1)) then redis.call('hincrby', KEYS[1], ARGV[2], 1); redis.call('pexpire', KEYS[1], ARGV[1]); return nil; end; return redis.call('pttl', KEYS[1]);" "1" "redisson_sortedset_lock:{INBOUND}" "30000" "acc3f971-3a53-42a2-b47d-37de483f9e51:46"
1720591768.803373 [0 lua] "exists" "redisson_sortedset_lock:{INBOUND}"
1720591768.803377 [0 lua] "hexists" "redisson_sortedset_lock:{INBOUND}" "acc3f971-3a53-42a2-b47d-37de483f9e51:46"
1720591768.803381 [0 lua] "hincrby" "redisson_sortedset_lock:{INBOUND}" "acc3f971-3a53-42a2-b47d-37de483f9e51:46" "1"
1720591768.803386 [0 lua] "pexpire" "redisson_sortedset_lock:{INBOUND}" "30000"
1720591768.806063 [0 172.17.0.1:61092] "LPOP" "INBOUND"
1720591768.811666 [0 172.17.0.1:61114] "EVAL" "local val = redis.call('get', KEYS[3]); if val ~= false then return tonumber(val);end; if (redis.call('hexists', KEYS[1], ARGV[3]) == 0) then return nil;end; local counter = redis.call('hincrby', KEYS[1], ARGV[3], -1); if (counter > 0) then redis.call('pexpire', KEYS[1], ARGV[2]); redis.call('set', KEYS[3], 0, 'px', ARGV[5]); return 0; else redis.call('del', KEYS[1]); redis.call(ARGV[4], KEYS[2], ARGV[1]); redis.call('set', KEYS[3], 1, 'px', ARGV[5]); return 1; end; " "3" "redisson_sortedset_lock:{INBOUND}" "redisson_lock__channel:redisson_sortedset_lock:{INBOUND}" "redisson_unlock_latch:redisson_sortedset_lock:{INBOUND}:2882dc5cc947b2ef1a7d0cbf3feef717" "0" "30000" "acc3f971-3a53-42a2-b47d-37de483f9e51:46" "PUBLISH" "13500"
1720591768.811749 [0 lua] "get" "redisson_unlock_latch:redisson_sortedset_lock:{INBOUND}:2882dc5cc947b2ef1a7d0cbf3feef717"
1720591768.811762 [0 lua] "hexists" "redisson_sortedset_lock:{INBOUND}" "acc3f971-3a53-42a2-b47d-37de483f9e51:46"
1720591768.811771 [0 lua] "hincrby" "redisson_sortedset_lock:{INBOUND}" "acc3f971-3a53-42a2-b47d-37de483f9e51:46" "-1"
1720591768.811778 [0 lua] "pexpire" "redisson_sortedset_lock:{INBOUND}" "30000"
1720591768.811788 [0 lua] "set" "redisson_unlock_latch:redisson_sortedset_lock:{INBOUND}:2882dc5cc947b2ef1a7d0cbf3feef717" "0" "px" "13500"
1720591768.814173 [0 172.17.0.1:61058] "DEL" "redisson_unlock_latch:redisson_sortedset_lock:{INBOUND}:2882dc5cc947b2ef1a7d0cbf3feef717"
1720591769.402131 [0 172.17.0.1:61084] "EVAL" "if ((redis.call('exists', KEYS[1]) == 0) or (redis.call('hexists', KEYS[1], ARGV[2]) == 1)) then redis.call('hincrby', KEYS[1], ARGV[2], 1); redis.call('pexpire', KEYS[1], ARGV[1]); return nil; end; return redis.call('pttl', KEYS[1]);" "1" "redisson_sortedset_lock:{INBOUND}" "30000" "acc3f971-3a53-42a2-b47d-37de483f9e51:46"
1720591769.402191 [0 lua] "exists" "redisson_sortedset_lock:{INBOUND}"
1720591769.402207 [0 lua] "hexists" "redisson_sortedset_lock:{INBOUND}" "acc3f971-3a53-42a2-b47d-37de483f9e51:46"
1720591769.402223 [0 lua] "hincrby" "redisson_sortedset_lock:{INBOUND}" "acc3f971-3a53-42a2-b47d-37de483f9e51:46" "1"
1720591769.402236 [0 lua] "pexpire" "redisson_sortedset_lock:{INBOUND}" "30000"
1720591769.405111 [0 172.17.0.1:61070] "LPOP" "INBOUND"
```

## The contents of the lock
```
127.0.0.1:6379> HGETALL "redisson_sortedset_lock:{INBOUND}"
1) "acc3f971-3a53-42a2-b47d-37de483f9e51:46"
2) "2"
```

## Steps to reproduce in local environment
Unfortunately, I was unable to simulate a real network issue locally, but I have reproduced the issue using the `CLIENT PAUSE` command on the Redis server. Here are the steps that I did to catch the infinite loop of acquiring/releasing:
1. Run the application in a debug mode
2. Set up break-point in like 108 of the `RedissonLock` class.
    `Long ttl = tryAcquire(-1, leaseTime, unit, threadId);`
3. Set up a break-point whenever in the code after putting the task in the priority queue.
4. Trigger putting a new task to the priority queue.
5. The debugging stops at line 108 of the `RedissonLock` class.
6. Execute the client pause command on the Redis server.
    `redis-cli CLIENT PAUSE 6000`
7. Resume the application.
8. The publisher thread stops working.
9. Repeat steps 4-7 a few times if need to catch the infinite loop of acquiring/releasing.

**Comment by mrniko** [comment]
> The publisher thread falls into an infinite loop of acquiring/releasing

Can you share the stacktrace?

**Comment by adr-wojtasik** [comment]
I already shared it in the section `Stack trace from the publisher thread`.

**Comment by mrniko** [comment]
Sorry I don't see an infinity loop in it.

**Comment by adr-wojtasik** [comment]
Please take a look at the above Redis server logs - you can see there that the counter is decremented and incremented indefinitely:
- `"hincrby" "redisson_sortedset_lock:{INBOUND}" "acc3f971-3a53-42a2-b47d-37de483f9e51:46" "-1"`
- `"hincrby" "redisson_sortedset_lock:{INBOUND}" "acc3f971-3a53-42a2-b47d-37de483f9e51:46" "1"`

The lock is never removed and the task is never published. The following commands are missing in the case of infinity loop of acquiring/releasing:
- `"del" "redisson_sortedset_lock:{INBOUND}"`
- `"PUBLISH" "redisson_lock__channel:redisson_sortedset_lock:{INBOUND}" ...`

**Comment by adr-wojtasik** [comment]
This is the implementation used to unlock the lock:
```
protected RFuture<Boolean> unlockInnerAsync(long threadId, String requestId, int timeout) {
        return evalWriteSyncedAsync(getRawName(), LongCodec.INSTANCE, RedisCommands.EVAL_BOOLEAN,
                              "local val = redis.call('get', KEYS[3]); " +
                                    "if val ~= false then " +
                                        "return tonumber(val);" +
                                    "end; " +

                                    "if (redis.call('hexists', KEYS[1], ARGV[3]) == 0) then " +
                                        "return nil;" +
                                    "end; " +
                                    "local counter = redis.call('hincrby', KEYS[1], ARGV[3], -1); " +
                                    "if (counter > 0) then " +
                                        "redis.call('pexpire', KEYS[1], ARGV[2]); " +
                                        "redis.call('set', KEYS[3], 0, 'px', ARGV[5]); " +
                                        "return 0; " +
                                    "else " +
                                        "redis.call('del', KEYS[1]); " +
                                        "redis.call(ARGV[4], KEYS[2], ARGV[1]); " +
                                        "redis.call('set', KEYS[3], 1, 'px', ARGV[5]); " +
                                        "return 1; " +
                                    "end; ",
                                Arrays.asList(getRawName(), getChannelName(), getUnlockLatchName(requestId)),
                                LockPubSub.UNLOCK_MESSAGE, internalLockLeaseTime,
                                getLockName(threadId), getSubscribeService().getPublishCommand(), timeout);
    }
```
The else block is never invoked here because the counter is always greater than 1.

**Comment by mrniko** [comment]
thank you for the details.

This case may happen if the same lock was acquired twice by the same thread.

**Comment by mrniko** [comment]
Can you share all RedisResponseTimeoutExceptions ?

**Comment by adr-wojtasik** [comment]
Sure, there were a total of **12** `RedisResponseTimeoutException` in the last 30 days. There are 2 kinds of stack trace that differ regarding a message that starts with "Check connection with Redis node: ...":
- "for TCP packet drops or bandwidth limits":
    ```
    org.redisson.client.RedisResponseTimeoutException: Redis server response timeout (3000 ms) occured after 0 retry attempts, is non-idempotent command: true Check connection with Redis node: <node_info> for TCP packet drops or bandwidth limits.  Try to increase nettyThreads and/or timeout settings. Command: (LPOP), params: [INBOUND], channel: [id: 0x8f7c13b6, <node_info>]
	    at org.redisson.command.RedisExecutor.lambda$scheduleResponseTimeout$11(RedisExecutor.java:447)
	    at io.netty.util.HashedWheelTimer$HashedWheelTimeout.expire(HashedWheelTimer.java:672)
	    at io.netty.util.HashedWheelTimer$HashedWheelBucket.expireTimeouts(HashedWheelTimer.java:747)
	    at io.netty.util.HashedWheelTimer$Worker.run(HashedWheelTimer.java:472)
	    at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	    at java.base/java.lang.Thread.run(Unknown Source)
    ```

- "for TCP packet drops":
    ```
    org.redisson.client.RedisResponseTimeoutException: Redis server response timeout (3000 ms) occured after 0 retry attempts, is non-idempotent command: true Check connection with Redis node: <node_info> for TCP packet drops.  Try to increase nettyThreads and/or timeout settings. Command: (LPOP), params: [INBOUND], channel: [id: 0x0b05ea34, <node_info>]
	    at org.redisson.command.RedisExecutor.lambda$scheduleResponseTimeout$8(RedisExecutor.java:387)
	    at io.netty.util.HashedWheelTimer$HashedWheelTimeout.expire(HashedWheelTimer.java:672)
	    at io.netty.util.HashedWheelTimer$HashedWheelBucket.expireTimeouts(HashedWheelTimer.java:747)
	    at io.netty.util.HashedWheelTimer$Worker.run(HashedWheelTimer.java:472)
	    at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	    at java.base/java.lang.Thread.run(Unknown Source)
    ```

Additionally, I have found **74** occurrences of the `RedisTimeoutException` in the last 30 days. Here is the stack trace which looks the same for each occurrence:
```
org.redisson.client.RedisTimeoutException: Command execution timeout for command: (PING), params: [], Redis client: [addr=<node_info>]
	at org.redisson.client.RedisConnection.lambda$async$0(RedisConnection.java:244)
	at io.netty.util.HashedWheelTimer$HashedWheelTimeout.expire(HashedWheelTimer.java:672)
	at io.netty.util.HashedWheelTimer$HashedWheelBucket.expireTimeouts(HashedWheelTimer.java:747)
	at io.netty.util.HashedWheelTimer$Worker.run(HashedWheelTimer.java:472)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Unknown Source)
```

**Comment by shiranyosef1** [comment]
any update on this? seeing similar issues

**Comment by mrniko** [comment]
@shiranyosef1 

Can you share any reproducer?

**Comment by seakider** [comment]
As described in the "Logs from the Redis server" section, the problem is indeed caused by the consumer's persistent lock holding. This has been fixed in version Redison-3.38.1. Can you try to verify it with the latest version? This problem should not occur.
@shiranyosef1

**Comment by mrniko** [comment]
I'm closing it. Let me know if it happens with 3.38.1 version or higher
